crates/chia-consensus/src/spendbundle_validation.rs

```diff
@@ -0,0 +1,400 @@
+            (AGG_SIG_ME, &spend.agg_sig_me),
+        ];
+        condition_items_pairs
+            .into_iter()
+            .flat_map(move |(condition, items)| {
+                let spend_clone = spend.clone();
+                items.iter().map(move |(pk, msg)| {
+                    (
+                        pk,
+                        make_aggsig_final_message(
+                            condition,
+                            msg.as_slice(),
+                            &spend_clone,
+                            constants,
+                        ),
+                    )
+                })
+            })
+    });
+    let unsafe_items = npcresult
+        .agg_sig_unsafe
+        .iter()
+        .map(|(pk, msg)| (pk, msg.as_slice().to_vec()));
+    let iter = iter.chain(unsafe_items);
+    // Verify aggregated signature
+    let (result, added) = cache
+        .lock()
+        .unwrap()
+        .aggregate_verify(iter, &spend_bundle.aggregated_signature);
+    if !result {
+        return Err(ErrorCode::BadAggregateSignature);
+    }
+    Ok((npcresult, added, start_time.elapsed()))
+}
+
+pub fn get_flags_for_height_and_constants(height: u32, constants: &ConsensusConstants) -> u32 {
+    let mut flags: u32 = 0;
+
+    if height >= constants.soft_fork4_height {
+        flags |= ENABLE_MESSAGE_CONDITIONS;
+    }
+
+    if height >= constants.soft_fork5_height {
+        flags |= DISALLOW_INFINITY_G1;
+    }
+
+    if height >= constants.hard_fork_height {
+        //  the hard-fork initiated with 2.0. To activate June 2024
+        //  * costs are ascribed to some unknown condition codes, to allow for
+        // soft-forking in new conditions with cost
+        //  * a new condition, SOFTFORK, is added which takes a first parameter to
+        //    specify its cost. This allows soft-forks similar to the softfork
+        //    operator
+        //  * BLS operators introduced in the soft-fork (behind the softfork
+        //    guard) are made available outside of the guard.
+        //  * division with negative numbers are allowed, and round toward
+        //    negative infinity
+        //  * AGG_SIG_* conditions are allowed to have unknown additional
+        //    arguments
+        //  * Allow the block generator to be serialized with the improved clvm
+        //   serialization format (with back-references)
+        flags = flags | ENABLE_BLS_OPS_OUTSIDE_GUARD | ENABLE_FIXED_DIV | ALLOW_BACKREFS;
+    }
+    flags
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::consensus_constants::TEST_CONSTANTS;
+    use crate::gen::condition_tools::u64_to_bytes;
+    use chia_bls::{sign, G2Element, SecretKey, Signature};
+    use chia_protocol::{Bytes, Bytes32};
+    use chia_protocol::{Coin, CoinSpend, Program};
+    use clvm_utils::tree_hash_atom;
+    use hex::FromHex;
+    use hex_literal::hex;
+    use std::sync::Arc;
+
+    #[test]
+    fn test_validate_no_pks() {
```

Comment:
there seems to be a lot of repetitions across these tests. did you try to factor something out or make it a parameterized test?